### DISCUSSION TITLE
Using FalkorDB as an application database (undo/redo)

### DISCUSSION THREAD

**Discussion by georgfaust** [FalkorDB/FalkorDB#54] [General]
Hello,

great to see that this fork exists!

I'm looking for a graph-database that I can use as an application database (create/update/query project data). One requirement I can't easily fulfill with "normal" graph databases is undo/redo.
For this I'd like to take a snapshot approach, as all other patterns are too complex.

Is it possible/feasible to just snapshot the complete redis-store on each change and restore the old state on "undo"? (I never used redis yet)

thanks for any input.

Sebastian

**Comment by gkorland** [comment]
Yes, this is a repeated request and on our priority to provide an MVCC support.
As for your question yes, Redis provides a[ Dump/Restore](https://redis.io/commands/dump/) API that lets you dump a specific key and load it back. Also Redis provides a [Copy](https://redis.io/commands/copy/) and [MOVE](https://redis.io/commands/move/) which you can use for temporary copy and then override the original key.

**Reply by georgfaust** [reply]
regarding MVCC. Do you propose an undo/redo implementation based on MVCC/timestamps rather than snapshots?

**Comment by georgfaust** [comment]
That's good to hear.
I think I'll go forward and build a POC then.
Do you know of any prior art (as this is a repeated request)?

**Comment by gkorland** [comment]
>  Do you know of any prior art (as this is a repeated request)?
For MVCC or Copy/Move?

**Reply by ahsf** [reply]
I meant usage of redisgraph/falkor as app-db with undo/redo.

So I just tried this out. If I understand correctly, as the graph is just one key, I just have to work on this key!? I can dump/restore the graph:

```
127.0.0.1:6379> GRAPH.QUERY g "CREATE (:L {foo: 1})"
...  

127.0.0.1:6379> GRAPH.QUERY g "MATCH (n:L) RETURN n"
...
  2) 1) 1) "foo"
        2) (integer) 1

127.0.0.1:6379> DUMP g
"\a\x81\x82...

127.0.0.1:6379> GRAPH.QUERY g "MATCH (n:L) SET n.foo=2 RETURN n"
...
  2) 1) 1) "foo"
        2) (integer) 2

127.0.0.1:6379> DEL g
(integer) 1

127.0.0.1:6379> RESTORE g 0 "\a\x81\x82...
OK

127.0.0.1:6379> GRAPH.QUERY g "MATCH (n:L) RETURN n"
...
  2) 1) 1) "foo"
        2) (integer) 1
```

COPY does not work, what am I doing wrong?

```
127.0.0.1:6379> COPY g save
(error) ERR not supported for this module key
```

There is also a "telemetry{<graph-name>}" key:

```
127.0.0.1:6379> KEYS *
1) "g"
2) "telemetry{g}"
```

do I have to worry about this when dumping/restoring?

**Reply by ahsf** [reply]
sorry, for the confusion. georgfaust is my private github account.

**Reply by gkorland** [reply]
https://github.com/FalkorDB/FalkorDB/issues/58

**Reply by georgfaust** [reply]
ok I'll go forward with using DUMP assuming I'll switch to COPY once its implemented.

**Comment by AviAvni** [comment]
can you describe more on your use case because I don't think that copying the database is a good solution

**Reply by georgfaust** [reply]
The Graph will hold application data (which is devices in rooms, connected by cables).
I need to implement undo/redo.
The easiest way to implement that is taking snapshots on each change.

gkorland mentioned that you'll implement MVCC which may offer another way to do that using timestamps. But I think it'll be way more complex.
And as the graph will not be really big, I don't think that any optimization is needed over snapshots.

**Reply by AviAvni** [reply]
Undo is just delete a node why snapshot?

**Reply by georgfaust** [reply]
one user action results in an arbitrary number of new / updated / deleted nodes and relations.

**Comment by howardshaw** [comment]
great to hear FalkorDB  will provide an MVCC support.Is there any progress?Thanks.
